Tests/StreamChatTests/Utils/CooldownTracker_Tests.swift

```diff
@@ -0,0 +1,109 @@
+//
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+@testable import StreamChat
+import StreamChatTestHelpers
+import XCTest
+
+final class CooldownTracker_Tests: XCTestCase {
+    var sut: CooldownTracker!
+    
+    func test_start_thenStartTimer() {
+        // GIVEN
+        let cooldownTime = 3
+        let timer = ScheduledStreamTimer_Mock()
+        sut = CooldownTracker(timer: timer)
+        
+        // WHEN
+        sut.start(with: cooldownTime)
+        
+        // THEN
+        XCTAssertEqual(timer.numberOfTimesStarted, 1)
+    }
+    
+    func test_start_whenCooldownIsZero_thenDontStartTimer() {
+        // GIVEN
+        let cooldownTime = 0
+        let timer = ScheduledStreamTimer_Mock()
+        sut = CooldownTracker(timer: timer)
+        
+        // WHEN
+        sut.start(with: cooldownTime)
+        
+        // THEN
+        XCTAssertEqual(timer.numberOfTimesStarted, 0)
+    }
+    
+    func test_start_whenDurationChanges_thenOnChangeIsCalled() {
+        // GIVEN
+        let cooldownTime = 3
+        let timer = ScheduledStreamTimer_Mock()
+        sut = CooldownTracker(timer: timer)
+        sut.onChange = { currentTime in
+            // THEN
+            XCTAssertEqual(cooldownTime, currentTime)
+        }
+        timer.onChange = { [weak self] in
+            self?.sut.onChange?(cooldownTime)
+        }
+        
+        // WHEN
+        sut.start(with: cooldownTime)
+    }
+    
+    func test_start_whenDurationChanges_andDurationIsNotZero_thenDecreaseDuration_andTimerIsNotStopped() {
+        // GIVEN
+        let cooldownTime = 3
+        let timer = ScheduledStreamTimer_Mock()
+        sut = CooldownTracker(timer: timer)
+        
+        // WHEN
+        sut.start(with: cooldownTime)
+        
+        // THEN
+        XCTAssertEqual(timer.numberOftimesStopped, 0)
+    }
```

Comment:
We are missing the `thenDecreaseDuration` assertion here. This test should also be written based on the suggestion above. Because here we need the onChange to be called twice, this is easily mockable with the expectation API, example:

```swift
func test_start_whenDurationChanges_andDurationIsNotZero_thenDecreaseDuration_andTimerIsNotStopped() {
        // GIVEN
        let cooldownTime = 3
        let timer = ScheduledStreamTimer_Mock()
        sut = CooldownTracker(timer: timer)

        let exp = expectation(description: "should decrease duration")
        exp.expectedFulfillmentCount = 2
        var durationChanged: Int = 3
        sut.onChange = { duration in
            durationChanged = duration
            exp.fulfill()
        }
        
        // WHEN
        sut.start(with: cooldownTime)
        timer.onChange?()
        timer.onChange?()
        
        // THEN
        waitForExpectations(timeout: 0.5)
        XCTAssertEqual(durationChanged, 2)
        XCTAssertEqual(timer.numberOftimesStopped, 0)
    }
```